### PR TITLE
Add additional functional functionality to Result.

### PR DESCRIPTION
**Pull request by jshier**
This PR adds `mapError` and `flatMapError` to transform `Result` error values; `withValue` and `withError` to execute closures with values without transforming; and `ifSuccess`, `ifFailure` to execute closures without values. This should cover most all of the `Result` use cases.

### CODE REVIEW

File: Source/Result.swift [hunk 1]
```diff
@@ -200,4 +200,101 @@ extension Result {
...
+                return try .failure(transform(error))
+            } catch {
+                return .failure(error)
+            }
+        case .success:
+            return self
+        }
+    }
+    
+    /// Evaluates the specified closure when the `Result` is a success, passing the unwrapped value as a paramter.
```
Comment (cnoon) [inline Source/Result.swift]: `paramter` -> `parameter`.

File: Source/Result.swift [hunk 2]
```diff
@@ -200,4 +200,101 @@ extension Result {
...
+    /// - Parameter closure: A closure that takes the success value of this instance.
+    /// - Returns: This `Result` instance, unmodified.
+    @discardableResult
+    public func withValue(_ closure: (Value) -> Void) -> Result {
+        if case let .success(value) = self { closure(value) }
+        
+        return self
+    }
+    
+    /// Evaluates the specified closure when the `Result` is a failure, passing the unwrapped error as a paramter.
```
Comment (cnoon) [inline Source/Result.swift]: `paramter` -> `parameter`.

File: Tests/ResultTests.swift [hunk 3]
```diff
@@ -277,4 +280,194 @@ class ResultTestCase: BaseTestCase {
             XCTFail("flatMap should preserve the failure error")
         }
     }
+    
+    // MARK: - map/flatMapError Tests
```
Comment (cnoon) [inline Tests/ResultTests.swift]: How about: `// MARK: - Error Mapping Tests`?

File: Tests/ResultTests.swift [hunk 4]
```diff
@@ -277,4 +280,194 @@ class ResultTestCase: BaseTestCase {
...
+        struct ResultError: Error {}
+        struct OtherError: Error {
+            let error: Error
+            init(error: Error) throws { throw ThrownError() }
+        }
+        struct ThrownError: Error {}
+        let result: Result<String> = .failure(ResultError())
+        
+        // When
+        let mappedResult = result.flatMapError { try OtherError(error: $0) }
```
Comment (cnoon) [inline Tests/ResultTests.swift]: Newline?

File: Tests/ResultTests.swift [hunk 5]
```diff
@@ -277,4 +280,194 @@ class ResultTestCase: BaseTestCase {
...
+        let result: Result<String> = .success("success")
+        var string = "success"
+        
+        // When
+        result.withError { string = "\(type(of: $0))" }
+        
+        // Then
+        XCTAssertEqual(string, "success")
+    }
+
```
Comment (cnoon) [inline Tests/ResultTests.swift]: How about: `// MARK: - If Success or Failure Tests`?

File: Tests/ResultTests.swift [hunk 6]
```diff
@@ -277,4 +280,194 @@ class ResultTestCase: BaseTestCase {
...
+        let mappedResult = result.flatMapError { try OtherError(error: $0) }
+        // Then
+        if let error = mappedResult.error {
+            XCTAssertTrue(error is ThrownError)
+        } else {
+            XCTFail("mapError should capture thrown error value")
+        }
+    }
+    
+    // MARK: with/if Tests
```
Comment (cnoon) [inline Tests/ResultTests.swift]: How about: `// MARK: - With Value or Error Tests`?

File: Tests/ResultTests.swift [hunk 7]
```diff
@@ -277,4 +280,194 @@ class ResultTestCase: BaseTestCase {
...
+        struct ResultError: Error {}
+        let result: Result<String> = .success("first")
+        var string = "first"
+        var success = false
+        
+        // When
+        let endResult = result.map { _ in "second" }
+                              .flatMap { _ in "third" }
+                              .withValue { if $0 == "third" { string = "fourth" } }
+                              .ifSuccess { success = true }
```
Comment (cnoon) [inline Tests/ResultTests.swift]: How about the following formatting to stay consistent with the rest of the library?

```swift
let endResult = result
    .map { _ in "second" }
    .flatMap { _ in "third" }
    .withValue { if $0 == "third" { string = "fourth" } }
    .ifSuccess { success = true }
```

File: Tests/ResultTests.swift [hunk 8]
```diff
@@ -277,4 +280,194 @@ class ResultTestCase: BaseTestCase {
...
+        let result: Result<String> = .success("success")
+        var string = "success"
+        
+        // When
+        result.ifFailure { string = "failure" }
+        
+        // Then
+        XCTAssertEqual(string, "success")
+    }
+
```
Comment (cnoon) [inline Tests/ResultTests.swift]: `// MARK: - Functional Chaining Tests`?

File: Source/Result.swift (hunk 1 above)
Comment (jshier) [inline Source/Result.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Source/Result.swift (hunk 2 above)
Comment (jshier) [inline Source/Result.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Tests/ResultTests.swift (hunk 3 above)
Comment (jshier) [inline Tests/ResultTests.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Tests/ResultTests.swift (hunk 4 above)
Comment (jshier) [inline Tests/ResultTests.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Tests/ResultTests.swift (hunk 6 above)
Comment (jshier) [inline Tests/ResultTests.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Tests/ResultTests.swift (hunk 5 above)
Comment (jshier) [inline Tests/ResultTests.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Tests/ResultTests.swift (hunk 8 above)
Comment (jshier) [inline Tests/ResultTests.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Tests/ResultTests.swift (hunk 7 above)
Comment (jshier) [inline Tests/ResultTests.swift]: Addressed in [9bf29](https://github.com/Alamofire/Alamofire/pull/2135/commits/131354b48086ea94f3b3d9303347d2f64079bf29).

File: Tests/ResultTests.swift [hunk 9]
```diff
@@ -277,4 +280,204 @@ class ResultTestCase: BaseTestCase {
...
+        
+        // When
+        result.withError { string = "\(type(of: $0))" }
+        
+        // Then
+        #if swift(>=3.2)
+        XCTAssertEqual(string, "ResultError #1")
+        #else
+        XCTAssertEqual(string, "(ResultError #1)")
+        #endif
```
Comment (cnoon) [inline Tests/ResultTests.swift]: Could we dedent the pound defines?

Comment (jshier) [inline Tests/ResultTests.swift]: Addressed in [9df24a](https://github.com/Alamofire/Alamofire/pull/2135/commits/d518296937b845bb71d15109b449d9bf3f9df24a).

**Pull request merged**